components/cdc/src/observer.rs

```diff
@@ -99,48 +95,34 @@ impl CdcObserver {
 impl Coprocessor for CdcObserver {}
 
 impl<E: KvEngine> CmdObserver<E> for CdcObserver {
-    fn on_prepare_for_apply(&self, cdc: &ObserveHandle, rts: &ObserveHandle, region_id: u64) {
-        let is_observing = cdc.is_observing();
-        *self.last_batch_observing.borrow_mut() = is_observing;
-        if !is_observing {
+    // `CdcObserver::on_flush_applied_cmd_batch` should only invoke if `cmd_batches` is not empty
+    fn on_flush_applied_cmd_batch(&self, cmd_batches: &mut Vec<CmdBatch>, engine: &E) {
+        assert!(!cmd_batches.is_empty());
+        if cmd_batches.iter().all(|cb| cb.level != ObserveLevel::All) {
```

Comment:
This is redundant, we have a filter below.

Reply:
It is a fast path to avoid following action  (i.e making an empty vec) for most peers, see https://github.com/tikv/tikv/pull/10245#discussion_r642294846

Reply:
I don't think it is fast because it would iterate over all batches...

Reply:
Addressed, PTAL again.